src/lambda/utilities/getMainSourceFile.ts

```diff
@@ -146,3 +148,21 @@ async function getPythonSourceFileUri({
 
     throw new Error(`Python file expected at ${basePath}.py, but no file was found`)
 }
+
+async function getDotnetSourceFileUri({
+    fileExists,
+    root,
+    resource,
+}: Pick<OpenMainSourceFileUriContext, 'fileExists'> & {
+    root: vscode.Uri,
+    resource: CloudFormation.Resource
+}): Promise<vscode.Uri> {
+    const handler = resource.Properties!.Handler
+    const [packageName] = handler.split('::')
+    const handlerFilePath = path.join(root.fsPath, 'src', packageName, 'Program.cs')
```

Comment:
I feel like this should be the "fallback" path, if no other file can be dynamically determined.
Given that codeuri can essentially point at a zip file "anywhere", all we have to work with is the handler. @mpiroc is it reasonable to try to pull the class name from the handler and try looking for a file with the same name? Failing this, it could then "fall back" to looking for Program.cs.

Reply:
Correct me if I'm wrong but this is only used when creating a new SAM app and in this case the file name, code directory, namespace and handler are constant; i.e. the same thing is created every time SAM creates a Dotnet app. If is truly a constant, the code is already more complicated than need be and there is little point in spending more effort to further complicate it to support other "what if the code is elsewhere" scenarios.

It is apparent by the directory structure that there are broader objectives for `getMainSourceFileUri` but it is not used in broader scenarios and thus could be represented by an internal function within `createNewSamApp.ts`

Reply:
The SAM CLI is interchangeable, and there is no guarantee to the application it will produce. Otherwise it would make sense hardcode the expected file.

Reply:
1. Can you explain how the SAM CLI is interchangeable? I don't understand. I'm pretty confident you don't mean someone could roll their own SAM CLI compatible CLI.
2. Are you aware of any case in which the SAM CLI's `sam init` has generated a different handler for a specified runtime than it previously did?

It seems that a map along with an integration test that tests `sam init` for each runtime and verifies that the expected handler file exists would be much simpler to understand and maintain, more efficient, and no less robust than a case by case function for each runtime where the DotNet runtime considers all cases theoretically possible including with namespaces and without, and if a namespace is used, parsing each file in the directory associated with the namespace and attempting to identify the function corresponding to the handler in the SAM template.

To be specific. If a user or our extension calls `sam init --runtime dotnetcore2.1`, SAM _will_ create a file named `Program.cs` in `${samProjectDir}/src/HelloWorld` every time. We know how the movie ends, we don't need to run AI to consider all of the potential outcomes that could have been and prepare for each outcome, we can just deal with the known outcome. Or more specifically `sam init` can be viewed as  a pure function that always results in the same output for a given combination of inputs.

Testing this assumption simply means (in pseudo code):
```
for(const runtime in allSamRuntimes) {
   await exec(`sam init --runtime ${runtime}`)
   const relativeHandlerPath = relativeHandlerPathForRuntimes[runtime]
   const expectedHandlerExists = await fileExists(path.resolve(samProjectDir, 
relativHandlerPath)
   assert(expectedHandlerExists)
}
```

Yes, perhaps someday SAM could release a version of the CLI in which they decided to shuffle things around in which case our tests would fail against that SAM release and we'd need to modify our code check against a list of the known handler paths for a given runtime which is still simpler than the alternative.

Reply:
If the team does decide to  continue to do things the more complicated way, we should consider leveraging the same logic as `getLambdaHandlerCandidates` that is being developed in the CR to support DotNet Run/Debug CodeLens. Basically we might list the `.cs` files in the directory associated with the namespace specified in the SAM template (assuming a namespace is used), run `getLambdaHandlerCandidates` on each and open for editing the first matching `.cs` file containing the function name specified in the SAM template.

Reply:
But do keep in mind that no-one has yet to identify any situation in which the code in this CR does not work as intended so in some sense it appears any further changes are in the name of idealogical purity.... just another way of accomplishing the same thing. Unlike the run/debug code lens, during `sam init` we don't need to worry about every possible way a user could configure their template and code. SAM consistently initializes the way it initializes.

Reply:
@micholas As an example, SAM CLI's .NET template currently creates a file called `Program.cs` that contains a class called `Function`, which breaks two .NET conventions: 1. Filename should match the class contained in the file; 2. `Program.cs` should contain the `Main` method, but the project generated by `sam init` is a class library (i.e., no `Main` method).

If SAM CLI changes the filename, do we want to force users to upgrade the toolkit? We only bump the minimum SAM version based on breaking changes, and a change to a template shouldn't be breaking.

Reply:
@mpiroc To answer your question:

> If SAM CLI changes the filename, do we want to force users to upgrade the toolkit?

I believe the toolkit is already written such that it will warn users if they are using a version of the SAM CLI that is older or newer than was supported at the time the toolkit was generated. There are a host of reasons that it would likely be preferable for us to have users update the toolkit when a new version of SAM is released given the tightly coupled nature of the two. The unlikely chance that the SAM team would decide to generate completely different artifacts for previously supported runtimes could be just one more reason to synchronize SAM updates with toolkit updates.

I'm aware of the complexity of trying to cover every use case of trying to figure out where the handler generated by `sam init` might someday be or could have been (but isn't). I'm also very familiar with the handler functions that are in fact created. If I had implemented this for Java we wouldn't be having this conversation because both the directory and file name could be directly derived from the SAM template and I would have done so.

When `sam init` generates a DotNet project, the generated SAM template contains no reference to the actual name of the handler file.  This is why I brought up the idea of whether it is worth the effort of trying to algorithmically determine something we know in advance or whether we should consider simply using a map. **R2** question 2 would be one question we should definitely answer before choosing this simpler path. We may also consider asking the SAM team whether they have changed the path to a handler for a specific handler runtime, whether they intend to do so and what the likelihood is that they ever would. Given the response from most of our team members it appears the simplified approach is unpalatable regardless of the answer to these questions.

_Again if completely opposed to the idea of the simplified map approach, the other solution I mentioned in **R3** seems like the preferred option: inspecting all of the C# files in the directory associated with the handler namespace to determine which if any of them contain a function that might be the intended handler._